File: C/mapGenerator.c
#include <malloc.h>
#include <time.h>
#include <sys/stat.h>

#include "loadingBar.h"
#include "map.h"
#include "mapGenerator.h"

//? For positive integers only. Should not be an issue here.
int gcd(int a, int b)
{
    if (a == 0 || b == 0)
    {
        printf("%sERROR : can't find the greatest common divisor of 0 !%s\n", RED_COLOR, DEFAULT_COLOR);
        return 0;
    }

    while (a != b)
    {
        if (a > b)
        {
            a -= b;
        }
        else
        {
            b -= a;
        }
    }

    return a;
}



int lcm(int a, int b)
{
    if (a == 0 || b == 0)
    {
        return 0;
    }

    int c = gcd(a, b);

    return (a * b) / c;
}



int lcmOfArray(int nb, int array[nb])
{
    int common = 0;

    if (array != NULL && nb > 1)
    {
        common = lcm(array[0], array[1]);

        for (int i = 2; i < nb; i++)
        {
            common = lcm(common, array[i]);
        }
    }

    return common;
}






double* getCompleteMapSeaValue(completeMap* completeMap, int width_idx, int height_idx)
{
    double* sea_value = NULL;

    if (completeMap != NULL)
    {
        int width = completeMap->width;
        int height = completeMap->height;

        if (width_idx < 0 || width_idx >= width)
        {
            printf("%sERROR : invalid width_idx = %d when reading complete map sea value. Should be in range [0, %d]%s\n", RED_COLOR, width_idx, width - 1, DEFAULT_COLOR);
            return NULL;
        }
        
        if (height_idx < 0 || height_idx >= height)
        {
            printf("%sERROR : invalid height_idx = %d when reading complete map sea value. Should be in range [0, %d]%s\n", RED_COLOR, height_idx, height - 1, DEFAULT_COLOR);
            return NULL;
        }

        sea_value = (completeMap->sea_values) + height_idx * width + width_idx;
    }

    return sea_value;
}



color* getCompleteMapColor(completeMap* completeMap, int width_idx, int height_idx)
{
    color* map_color = NULL;

    if (completeMap != NULL)
    {
        int width = completeMap->width;
        int height = completeMap->height;

        if (width_idx < 0 || width_idx >= width)
        {
            printf("%sERROR : invalid width_idx = %d when reading complete map color. Should be in range [0, %d]%s\n", RED_COLOR, width_idx, width - 1, DEFAULT_COLOR);
            return NULL;
        }
        
        if (height_idx < 0 || height_idx >= height)
        {
            printf("%sERROR : invalid height_idx = %d when reading complete map color. Should be in range [0, %d]%s\n", RED_COLOR, height_idx, height - 1, DEFAULT_COLOR);
            return NULL;
        }

        map_color = completeMap->color_map[height_idx * width + width_idx];
    }

    return map_color;
}



color* colorize(double value, double sea_level, double min_value, double max_value)
{
    color* new_color = calloc(1, sizeof(color));

    int i = (int) ((value - min_value) * 255. / (max_value - min_value));
    int s = (int) ((value - min_value) * 200. / (sea_level - min_value));

    //? Show red dots on zero values.
    if (value == 0)
    {
        new_color->red_int = 255;
        new_color->green_int = 0;
        new_color->blue_int = 0;

        new_color->red_double = 1.;
        new_color->green_double = 0;
        new_color->blue_double = 0;
    }
    else if (value <= sea_level)
    {
        new_color->red_int = 50;
        new_color->green_int = 50;
        new_color->blue_int = s;

        new_color->red_double = 1./255 * 50;
        new_color->green_double = 1./255 * 50;
        new_color->blue_double = 1./255 * s;
    }
    else
    {
        new_color->red_int = 50;
        new_color->green_int = i;
        new_color->blue_int = 50;

        new_color->red_double = 1./255 * 50;
        new_color->green_double = 1./255 * i;
        new_color->blue_double = 1./255 * 50;
    }

    return new_color;
}





color** generateColorMap(map* map, double sea_level, unsigned int display_loading)
{
    clock_t start_time = clock();

    color** color_map = NULL;

    if (map != NULL)
    {
        int width = map->map_width * map->chunk_width;
        int height = map->map_height * map->chunk_height;

        color_map = calloc(width * height, sizeof(color*));
        if (color_map == NULL)
        {
            printf("%sColor map allocation was not successful. Returning NULL now...%s\n", RED_COLOR, DEFAULT_COLOR);
            return NULL;
        }

        double min_value = *getMapValue(map, 0, 0);
        double max_value = *getMapValue(map, 0, 0);

        // Getting min and max values.


        for (int i = 0; i < height; i++)
        {
            for (int j = 0; j < width; j++)
            {
                double map_value = *getMapValue(map, j, i);

                if (map_value < min_value)
                {
                    min_value = map_value;
                }

                if (map_value > max_value)
                {
                    max_value = map_value;
                }

                if (display_loading != 0)
                {
                    int nb_indents = display_loading - 1;

                    char base_str[100] = "Generating color map...            ";

                    predefined_loading_bar(i * width + j, 2 * height * width - 2, NUMBER_OF_SEGMENTS, base_str, nb_indents, start_time);
                }
            }
        }


        // Generating colors

        for (int i = 0; i < height; i++)
        {
            for (int j = 0; j < width; j++)
            {
                double map_value = *getMapValue(map, j, i);

                color* c = colorize(map_value, sea_level, min_value, max_value);
                
                color_map[i * width + j] = c;

                if (display_loading != 0)
                {
                    int nb_indents = display_loading - 1;

                    char base_str[100] = "Generating color map...            ";

                    predefined_loading_bar(height * width - 1 + i * width + j, 2 * height * width - 2, NUMBER_OF_SEGMENTS, base_str, nb_indents, start_time);
                }
            }
        }
    }

    return color_map;
}



double* setSeaLevel(map* map, double sea_level, unsigned int display_loading)
{
    clock_t start_time = clock();

    double* sea_map = NULL;

    if (map != NULL)
    {
        int width = map->map_width * map->chunk_width;
        int height = map->map_height * map->chunk_height;

        sea_map = calloc(width * height, sizeof(double));
        if (sea_map == NULL)
        {
            printf("%sSea map allocation was not successful. Returning NULL now...%s\n", RED_COLOR, DEFAULT_COLOR);
            return NULL;
        }

        for (int i = 0; i < height; i++)
        {
            for (int j = 0; j < width; j++)
            {
                double map_value = *getMapValue(map, j, i);

                if (map_value <= sea_level)
                {
                    sea_map[i * width + j] = sea_level;
                }
                else
                {
                    sea_map[i * width + j] = map_value;
                }


                if (display_loading != 0)
                {
                    int nb_indents = display_loading - 1;

                    char base_str[100] = "Setting sea level...               ";

                    predefined_loading_bar(i * width + j, height * width - 1, NUMBER_OF_SEGMENTS, base_str, nb_indents, start_time);
                }
            }
        }
    }

    return sea_map;
}





completeMap* newCompleteMapFromMap(map* map, double sea_level, unsigned int display_loading)
{
    completeMap* complete_map = NULL;

    if (map != NULL)
    {
        complete_map = calloc(1, sizeof(completeMap));

        int width = map->map_width * map->chunk_width;
        int height = map->map_height * map->chunk_height;

        complete_map->map = map;
        complete_map->width = width;
        complete_map->height = height;

        complete_map->sea_level = sea_level;

        
        complete_map->sea_values = setSeaLevel(map, sea_level, display_loading);

        if (display_loading != 0)
        {
            int nb_indents = display_loading - 1;
            indent_print(nb_indents, "\n");
        }

        complete_map->color_map = generateColorMap(map, sea_level, display_loading);
    }

    return complete_map;
}



completeMap* newCompleteMap(int number_of_layers, int gradGrids_width[number_of_layers], int gradGrids_height[number_of_layers],
                            int size_factors[number_of_layers], double layers_factors[number_of_layers],
                            int map_width, int map_height, double sea_level, unsigned int display_loading)
{
    map* map = newMap(number_of_layers, gradGrids_width, gradGrids_height, size_factors, layers_factors, map_width, map_height, display_loading);

    completeMap* complete_map = NULL;

    if (map != NULL)
    {
        complete_map = newCompleteMapFromMap(map, sea_level, display_loading);
    }

    return complete_map;
}





//? Generate square chunks with automatic size factors.
map* get2dMap(int number_of_layers, int gradGrids_dimension[number_of_layers], double layers_factors[number_of_layers],
                    int map_width, int map_height, unsigned int display_loading)
{
    clock_t start_time = clock();

    int m_loading = display_loading;

    if (display_loading != 0)
    {
        int nb_indents = display_loading - 1;
        indent_print(nb_indents, "Generating a 2d map...\n");
        m_loading += 1;
    }

    int lcm = lcmOfArray(number_of_layers, gradGrids_dimension);

    int size_factors[number_of_layers];

    // size factors should match gradient grids dimensions - 1
    int gradGrid_corresponding_dimensions[number_of_layers];

    for (int i = 0; i < number_of_layers; i++)
    {
        size_factors[i] = lcm / gradGrids_dimension[i];
        gradGrid_corresponding_dimensions[i] = gradGrids_dimension[i] + 1;
    }


    map* new_map = newMap(number_of_layers, gradGrid_corresponding_dimensions, gradGrid_corresponding_dimensions, size_factors, layers_factors,
                                map_width, map_height, m_loading);


    if (display_loading == 1)
    {
        double total_time = (double) (clock() - start_time)/CLOCKS_PER_SEC;
        char final_string[200] = "";

        snprintf(final_string, sizeof(final_string), "%sSUCCESS :%s The 2d map generation took a total of %.4lf second(s).\n",
                                GREEN_COLOR, DEFAULT_COLOR, total_time);
        
        int nb_indents = display_loading - 1;
        indent_print(nb_indents, final_string);
    }

    return new_map;
}



//? Generate square chunks with automatic size factors and creates sea and color maps.
completeMap* fullGen(int number_of_layers, int gradGrids_dimension[number_of_layers], double layers_factors[number_of_layers],
                         int map_width, int map_height, double sea_level, unsigned int display_loading)
{
    clock_t start_time = clock();
    int m_loading = display_loading;

    if (display_loading != 0)
    {
        int nb_indents = display_loading - 1;
        indent_print(nb_indents, "Generating a complete map...\n");
        m_loading += 1;
    }

    map* new_map = get2dMap(number_of_layers, gradGrids_dimension, layers_factors, map_width, map_height, m_loading);

    if (display_loading != 0)
    {
        int nb_indents = display_loading;
        indent_print(nb_indents, "\n");
    }

    completeMap* new_complete_map = newCompleteMapFromMap(new_map, sea_level, m_loading);

    if (display_loading == 1)
    {
        double total_time = (double) (clock() - start_time)/CLOCKS_PER_SEC;
        char final_string[200] = "";

        snprintf(final_string, sizeof(final_string), "%sSUCCESS :%s The complete map generation took a total of %.4lf second(s).\n",
                                GREEN_COLOR, DEFAULT_COLOR, total_time);
        
        int nb_indents = display_loading - 1;
        indent_print(nb_indents, final_string);
    }

    return new_complete_map;
}





void writeSeaMapFile(completeMap* completeMap, char path[])
{
    FILE* f = NULL;

    f = fopen(path, "w");

    if (f != NULL)
    {
        fprintf(f, "Sea Map\n");

        int width = completeMap->width;
        int height = completeMap->height;

        double sea_level = completeMap->sea_level;

        // Writing the parameters
        fprintf(f, "width=%d\nheight=%d\nsea_level=% .8lf\n", width, height, sea_level);

        map* current_map = completeMap->map;
        int map_width = current_map->map_width;
        int map_height = current_map->map_height;

        fprintf(f, "map_width_in_chunks=%d\nmap_height_in_chunks=%d\n", map_width, map_height);

        // Writing the values
        for (int i = 0; i < height; i++)
        {
            for (int j = 0; j < width; j++)
            {
                double value = *getCompleteMapSeaValue(completeMap, j, i);

                if (j != width - 1)
                {
                    fprintf(f, "% .8lf\t", value);
                }
                else
                {
                    fprintf(f, "% .8lf\n", value);
                }
            }
        }

        fclose(f);
    }
    else
    {
        printf("%sERROR : could not open file in writing mode at path '%s'%s\n", RED_COLOR, path, DEFAULT_COLOR);
        return;
    }
}



void writeColorIntMapFile(color** color_map, int width, int height, char path[])
{
    FILE* f = NULL;

    f = fopen(path, "w");

    if (f != NULL)
    {
        fprintf(f, "Color Int Map\n");

        // Writing the parameters
        fprintf(f, "width=%d\nheight=%d\n", width, height);

        // Writing the vectors
        for (int i = 0; i < height; i++)
        {
            for (int j = 0; j < width; j++)
            {
                color* color = color_map[j + i * width];

                if (j != width - 1)
                {
                    fprintf(f, "(%d,%d,%d)\t", color->red_int, color->green_int, color->blue_int);
                }
                else
                {
                    fprintf(f, "(%d,%d,%d)\n", color->red_int, color->green_int, color->blue_int);
                }
            }
        }

        fclose(f);
    }
    else
    {
        printf("%sERROR : could not open file in writing mode at path '%s'%s\n", RED_COLOR, path, DEFAULT_COLOR);
        return;
    }
}



void writeColorDoubleMapFile(color** color_map, int width, int height, char path[])
{
    FILE* f = NULL;

    f = fopen(path, "w");

    if (f != NULL)
    {
        fprintf(f, "Color Double Map\n");

        // Writing the parameters
        fprintf(f, "width=%d\nheight=%d\n", width, height);

        // Writing the vectors
        for (int i = 0; i < height; i++)
        {
            for (int j = 0; j < width; j++)
            {
                color* color = color_map[j + i * width];

                if (j != width - 1)
                {
                    fprintf(f, "(%.8lf,%.8lf,%.8lf)\t", color->red_double, color->green_double, color->blue_double);
                }
                else
                {
                    fprintf(f, "(%.8lf,%.8lf,%.8lf)\n", color->red_double, color->green_double, color->blue_double);
                }
            }
        }

        fclose(f);
    }
    else
    {
        printf("%sERROR : could not open file in writing mode at path '%s'%s\n", RED_COLOR, path, DEFAULT_COLOR);
        return;
    }
}



void writeCompleteMapFiles(completeMap* complete_map, char folder_path[])
{
    //! TEMPORARY
    struct stat st = {0};

    if (stat(folder_path, &st) == -1)
    {
        // Should check what 0700 permissions are exactly.
        mkdir(folder_path, 0700);
    }

    char sea_map_path[200] = "";
    snprintf(sea_map_path, sizeof(sea_map_path), "%ssea_map.txt", folder_path);
    writeSeaMapFile(complete_map, sea_map_path);

    char color_int_path[200] = "";
    snprintf(color_int_path, sizeof(color_int_path), "%scolor_int_map.txt", folder_path);
    writeColorIntMapFile(complete_map->color_map, complete_map->width, complete_map->height, color_int_path);

    char color_double_path[200] = "";
    snprintf(color_double_path, sizeof(color_double_path), "%scolor_double_map.txt", folder_path);
    writeColorDoubleMapFile(complete_map->color_map, complete_map->width, complete_map->height, color_double_path);
}



completeMap* readCompleteMapFiles(char path[]);





void freeCompleteMap(completeMap* completeMap)
{
    if (completeMap != NULL)
    {
        if (completeMap->sea_values != NULL)
        {
            free(completeMap->sea_values);
        }

        if (completeMap->color_map != NULL)
        {
            for (int i = 0; i < completeMap->height; i++)
            {
                for (int j = 0; j < completeMap->width; j++)
                {
                    free(getCompleteMapColor(completeMap, j, i));
                }
            }

            free(completeMap->color_map);
        }


        freeMap(completeMap->map);


        free(completeMap);        
    }
}

File: Python/mapGenerator.py

#? To authorize type hints of class itself (ie Map here)
#? answer found here : https://stackoverflow.com/questions/33533148/how-do-i-type-hint-a-method-with-the-type-of-the-enclosing-class
#from typing import Self                    #? Python 3.11+
from __future__ import annotations          #? Python 3.7+

import random as rd
import numpy as np
import matplotlib.pyplot as plt
from gradientGrid import GradientGrid
from layer import Layer
from chunk import Chunk
from map import Map
import interpreter as interp

class CompleteMap:
    """
    CompleteMap :
    -------------
    The CompleteMap class made out of a Map for enhanced processing.
    """
    
    
    #? -------------------------- Static ------------------------- #
    
    COMPLETE_MAP_ENCODING = b'\x04'
    
    @staticmethod
    def colorize(value: float, sea_level: float, min_value: float, max_value: float) -> tuple[float]:
        """Returns the color associated with the given value for the given parameters. The color is a
        tuple of 3 floats between `0.` and `1.`.

        Args:
            `value` (float): altitude value to be colorized.
            `sea_level` (float): altitude of the sea. Below this point, an ocean color will be applied. Above, a land color will be applied.
            `min_value` (float): The minimum altitude value of the whole map. Used to determine the brightness of the color.
            `max_value` (float): The maximum altitude value of the whole map. Used to determine the brightness of the color.

        Returns:
            tuple[int]: the color associated to the given altitude value
        
        Note:
            Returns a red color `(1., 0., 0.)` if the altitude value is exactly 0.. Used to spot chunks.
        """
        
        #? Show red dots on zero values. Could be removed.
        if value == 0:
            return (255, 0, 0)
        elif value <= sea_level:
            blue = int((value - min_value) * 200. / (sea_level - min_value))
            o = 50
            return (o, o, blue)
        else:
            green = int((value - min_value) * 255. / (max_value - min_value))
            o = 50
            return (o, green, o)
    
    
    
    
    
    @staticmethod
    def generateCompleteMapFromScratch(grid_widths: list[int], grid_heights: list[int], size_factors: list[int], layers_factors: list[float],
                                        map_width: int, map_height: int, sea_level: float) -> CompleteMap:
        """Generates a new Map structure and every required pre-structure based on the given parameters.

        Args:
            `grid_widths` (list[int]): the list of width to be used to generate the GradientGrids of each Chunk.
            `grid_heights` (list[int]): the list of height to be used to generate the GradientGrids of each Chunk.
            `size_factors` (list[int]): the list of size factors to be used to generate the Layers of each Chunk.
            `layers_factors` (list[float]): the list of layer factors to be used to generate each Chunk.
            `map_width` (int): the width of the map in chunks.
            `map_height` (int): the height of the map in chunks.
            `sea_level` (float): the altitude of the sea in the final CompleteMap.

        Returns:
            CompleteMap: the newly generated CompleteMap structure.
        
        Note:
            Every list should have the same length, which is the number of layers per Chunk.
            The sum of layer factors should not be 0.
            The given gradient grid dimensions and size factors should match so that the product `grid_dim * size_factor` is a constant Vector2D.
        """
        
        #* -------------- Verifications : Begin --------------
        
        #TODO
        
        #* -------------- Verifications : End   --------------
        
        
        
        new_map = Map.generateMapFromScratch(grid_widths=grid_widths, grid_heights=grid_heights, size_factors=size_factors, layers_factors=layers_factors,
                                             map_width=map_width, map_height=map_height)
        
        return CompleteMap(new_map, sea_level)
    
    
    
    @staticmethod
    def write(complete_map: CompleteMap, path: str=None, append: bool=False) -> bytes:
        bytes_str : bytes = b''
        bytes_str += CompleteMap.COMPLETE_MAP_ENCODING
        bytes_str += Map.write(complete_map.map)
        bytes_str += interp.bytesNumber(complete_map.sea_level)
        height = map.map_height * map.chunk_height
        width = map.map_width * map.chunk_width
        for i in range(height):
            for j in range(width):
                bytes_str += interp.bytesNumber(complete_map.sea_values[i,j])
                bytes_str += interp.bytesNumber(complete_map.color_map[i,j,0])
                bytes_str += interp.bytesNumber(complete_map.color_map[i,j,1])
                bytes_str += interp.bytesNumber(complete_map.color_map[i,j,2])
        if path!=None:
            if append: f=open(path,"ab")
            else: f=open(path,"wb")
            f.write(bytes_str)
        return bytes_str
    
    
    
    @staticmethod
    def read(path: str) -> CompleteMap:
        #TODO
        return None
    
    
    #? ------------------------ Instances ------------------------ #
    
    def __init__(self, map: Map, sea_level: float) -> None:
        """Initialize a CompleteMap structure based on the given Map and sea altitude.

        Args:
            map (Map): the map structure to build the complete map from.
            sea_level (float): the altitude of the sea.
        """
        
        self.map = None
        self.sea_level = 0
        
        self.sea_values = None
        self.color_map = None
        
        if type(map) != Map or float(sea_level) != sea_level:
            return
        
        self.map = map
        self.sea_level = sea_level
        
        alt = self.map.getFullMap()
        
        if type(alt) == np.ndarray:
            pass
        else:
            alt = np.array(alt)
        self.sea_values = np.where(alt >= sea_level, alt, sea_level)
        
        self.generateColorMap()
    
    
    
    def generateColorMap(self) -> None:
        """Initialize and generate the color_map array of this complete map object.
        """
        
        width = self.map.map_width * self.map.chunk_width
        height = self.map.map_height * self.map.chunk_height
        
        alt = self.map.getFullMap()
        
        if type(alt) == np.ndarray:
            pass
        else:
            alt = np.array(alt)
        
        max_alt = np.max(alt)
        min_alt = np.min(alt)
        
        color_map = [[(0, 0, 0) for j in range(width)] for i in range(height)]
        
        for i in range(height):
            for j in range(width):
                color_map[i][j] = CompleteMap.colorize(alt[i][j], self.sea_level, min_alt, max_alt)
        
        self.color_map = np.array(color_map, dtype=np.uint8)
        
        
        
        


class MapGenerator:
    """
    MapGenerator :
    -------------
    The static MapGenerator class used to process Maps, generate CompleteMaps, and display.
    """
    
    
    #? -------------------------- Static ------------------------- #
    
    @staticmethod
    def gcd(a: int, b: int) -> int:
        """Finds the Greatest Common Divisor of the two given integers.
        Only process strictly positive integers.

        Args:
            `a` (int): the first integer.
            `b` (int): the second integer.

        Returns:
            int: the gcd of the two given arguments.
        """
        
        if a == 0 or b == 0:
            print("Error : Can't find the greatest common divisor of 0.")
            return 0
        
        if a < 0 or b < 0:
            print("Error : this algorithm supports positive integers only.")
            return
        
        while a != b:
            if a > b:
                a -= b
            else:
                b -= a
        
        return a
    
    
    @staticmethod
    def lcm(a: int, b: int) -> int:
        """Computes the Lowest Common Multiple from the two given integers.
        They shall be positive.

        Args:
            `a` (int): the first integer.
            `b` (int): the second integer.

        Returns:
            int: the computed lcm.
        """
        
        if a < 0 or b < 0:
            print("Error : this algorithm supports positive integers only.")
            return -1
        
        if a == 0 or b == 0:
            return 0
        
        c = MapGenerator.gcd(a, b)
        return (a * b) / c
        
    
    
    @staticmethod
    def lcmOfArray(array: list[int]) -> int:
        """Computes the Lowest Common Multiple of a list of positive integers.

        Args:
            `array` (list[int]): list of integers to compute the lcm from. Should be of length >= 2.

        Returns:
            int: final computed lcm of the list. Returns `-1` if negative integers were found.
        """
        
        common = -1
        
        l = len(array)
        if l >= 2:
            
            common = MapGenerator.lcm(array[0], array[1])
            
            if common == 0:
                return 0
            
            elif common == -1:
                print("Error : something went wrong in the lcm function.")
                return -1
            
            for i in range(2, l):
                
                common = MapGenerator.lcm(common, array[i])
                
                if common == 0:
                    return 0
                
                elif common == -1:
                    print("Error : something went wrong in the lcm function.")
                    return -1
        
        return common
    
    
    
    
    @staticmethod
    def getRandomSeed() -> int:
        """Gets a random valid integer seed to be used in the generation. Can be used this way :
        
        >>> seed = MapGenerator.getRandomSeed()
        
        >>> MapGenerator.setRandomSeed(seed)
        
        >>> # MapGenerator generation

        Returns:
            int: a random valid integer seed to be used.
        """
        
        return rd.randint(1, GradientGrid.MAX_INT_SEED)
    
    
    
    
    @staticmethod
    def setRandomSeed(seed: int | float | str | bytes | bytearray | None = getRandomSeed()) -> None:
        """Set the seed for the future random generations.

        Args:
            `seed` (int | float | str | bytes | bytearray | None): the seed to be used in the future random generations.
                                                                   Defaults to a random integer between `1` and `GradientGrid.MAX_INT_SEED`.
        
        Note:
            Calls the setRandomSeed static method from the `GradientGrid` class.
        """
        
        GradientGrid.setRandomSeed(seed)
    
    
    
    
    
    @staticmethod
    def get2DMap(grid_sizes: list[int], layers_factors: list[float], map_width: int, map_height: int) -> Map:
        """Generates a map structure with square chunks and automatic size factors from the given parameters.

        Args:
            `grid_sizes` (list[int]): the array of gradientGrid dimensions to be used to generate the random gradient grids.
            `layers_factors` (list[float]): the array of layers factors.
            `map_width` (int): number of chunks in width.
            `map_height` (int): number of chunks in height.

        Returns:
            Map: the final map.
        
        Note:
            The real gradient grids dimensions will be the passed ones + 1 to correctly match the passed dimensions' lcm to the final maps.
        """
        
        #* -------------- Verifications : Begin --------------
        
        #TODO
        
        #* -------------- Verifications : End   --------------
        
        lcm = MapGenerator.lcmOfArray(grid_sizes)
        
        size_factors = [int(lcm/gsi) for gsi in grid_sizes]
        final_grid_sizes = [gsi+1 for gsi in grid_sizes]
        
        new_map = Map.generateMapFromScratch(grid_widths=final_grid_sizes, grid_heights=final_grid_sizes, size_factors=size_factors,
                                             layers_factors=layers_factors, map_width=map_width, map_height=map_height)
        
        return new_map
    
    
    
    @staticmethod
    def fullGen(grid_sizes: list[int], layers_factors: list[float], map_width: int, map_height: int, sea_level: float) -> CompleteMap:
        """Generates a completeMap structure with square chunks, automatic size factors and the given sea altitude.
        It consists in generating a new map with the `get2dMap` function, and then creating a new completeMap with the `newCompleteMapFromMap` function.

        Args:
            `grid_sizes` (list[int]): the array of gradientGrid dimensions to be used to generate the random gradient grids.
            `layers_factors` (list[float]): the array of layers factors.
            `map_width` (int): number of chunks in width.
            `map_height` (int): number of chunks in height.
            `sea_level` (float): sea altitude to use.

        Returns:
            CompleteMap: the final complete map.
        
        Note:
            The real gradient grids dimensions will be the passed ones + 1 to correctly match the passed dimensions' lcm to the final maps.
        """
        
        #* -------------- Verifications : Begin --------------
        
        #TODO
        
        #* -------------- Verifications : End   --------------
        
        new_map = MapGenerator.get2DMap(grid_sizes=grid_sizes, layers_factors=layers_factors, map_width=map_width, map_height=map_height)
        
        new_complete_map = CompleteMap(map=new_map, sea_level=sea_level)
        
        return new_complete_map
    
    
    
    @staticmethod
    def plotCompleteMap(complete_map: CompleteMap) -> None:
        
        #? --- Get data ---
        
        grid_sizes = []
        
        first_chunk: Chunk = complete_map.map.chunks[0][0]
        layer_factors = []
        
        layers_list: list[Layer] = first_chunk.layers
        for layer in layers_list:
            gsize = (layer.grid.width, layer.grid.height)
            
            if gsize[0] == gsize[1]:
                grid_sizes.append(gsize[0])
            else:
                grid_sizes.append(gsize)
        
        layer_factors = first_chunk.layers_factors
        
        
        map_size = (complete_map.map.map_width, complete_map.map.map_height)
        
        chunk_size = (complete_map.map.chunk_width, complete_map.map.chunk_height)
        
        #? --- Plotting ---
        
        fig = plt.figure(figsize=(20, 10))
        fig.suptitle("SEED = " + str(GradientGrid.CURRENT_SEED) + "\nCompleteMap of sizes = " + str(map_size) \
            + "\nGrid sizes = " + str(grid_sizes) + " | Noise factors = " + str(layer_factors) \
            + " | Sea Level = " + str(complete_map.sea_level))
        
        # 2D Map
        ax2D = fig.add_subplot(1, 2, 1)
        ax2D.imshow(complete_map.color_map)
        
        # 3D Map
        ax3D = fig.add_subplot(1, 2, 2, projection='3d')
        
        x = np.linspace(0, map_size[0], chunk_size[0] * map_size[0])
        y = np.linspace(0, map_size[1], chunk_size[1] * map_size[1])
        x, y = np.meshgrid(x, y)
        
        
        if not type(complete_map.color_map) == np.ndarray:
            print("Color map should be an numpy array.")
        
        
        
        
        
        if np.issubdtype(complete_map.color_map.dtype, np.integer):
            cmap = complete_map.color_map.astype(np.float32)
            cmap /= 255
        elif not np.issubdtype(complete_map.color_map.dtype, np.floating):
            # WHAT ARE YOU TRYING TO DO ???
            print("Error : cmap can only be integers in [0, 255] or floating point values in [0., 1.]")
            print("You tried to use : ", complete_map.color_map.dtype)
            cmap = complete_map.color_map
        else:
            cmap = complete_map.color_map
        
        
        
        
        # print(np.shape(y), np.shape(x), np.shape(complete_map.sea_values), np.shape(complete_map.color_map))
        surf = ax3D.plot_surface(y, x, np.array(complete_map.sea_values), facecolors=np.array(cmap))
        
        xylim = max(map_size[0], map_size[1])
        ax3D.set_xlim(0, xylim)
        ax3D.set_ylim(0, xylim)
        
        zlim = min(map_size[0], map_size[1])
        ax3D.set_zlim(-zlim, zlim)
        
        plt.show(block=False)
        
        input("Press [Enter] to quit.")
    
    
    #? ------------------------ Instances ------------------------ #





# --------------------------------- # Testing MapGenerator # --------------------------------- #

if __name__ == "__main__":
    print("# --------------------------------- # Testing MapGenerator # --------------------------------- #\n")
    
    GradientGrid.setRandomSeed("seed")
    
    grid_dim = [4, 6]
    size_factors = [5, 3]
    layer_factors = [1, .1]
    
    
    map_dimensions = (2, 3)
    
    print("Generating a new map from scratch with parameters : ")
    print(" - map dimensions : ", map_dimensions)
    print(" - grids square dimensions : ", grid_dim)
    print(" - layers size_factors : ", size_factors)
    print(" - chunk layer factors : ", layer_factors)
    
    map = Map.generateMapFromScratch(grid_dim, grid_dim, size_factors, layer_factors, map_dimensions[0], map_dimensions[1])
    
    
    sea_level=0
    
    complete_map = CompleteMap(map, sea_level)
    
    
    MapGenerator.plotCompleteMap(complete_map)
    
    
    
    
    print("------------------------")
    print("Generating a new complete map from the `full gen` function with the given parameters :")
    
    grid_sizes = [2, 5, 9]
    lay_fac = [1, .1, .01]
    map_dim = (5, 7)        #? Width x Height
    sea_lvl = 0.2
    
    print(" - grid dimensions : ", grid_sizes)
    print("     ---> grid sizes will be used with a +1 for the layers size to be the lcm of the dimensions above : ",
          MapGenerator.lcmOfArray(grid_sizes))
    print(" - layer factors : ", lay_fac)
    print(" - map dimensions : ", map_dim)
    print(" - level of the sea : ", sea_lvl)
    
    other_complete_map = MapGenerator.fullGen(grid_sizes=grid_sizes, layers_factors=lay_fac,
                                              map_width=map_dim[0], map_height=map_dim[1],
                                              sea_level=sea_lvl)
    
    
    MapGenerator.plotCompleteMap(other_complete_map)
